base/src/store.rs

```diff
@@ -0,0 +1,62 @@
+use crate::error::BlockError;
+use core::future::Future;
+use core::pin::Pin;
+use cid::Cid;
+use std::path::Path;
+
+pub type StoreResult<T> = Pin<Box<dyn Future<Output = Result<T, BlockError>> + Send>>;
+
+/// Visibility of a block.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum Visibility {
+    /// Block is not announced on the network.
+    Private,
+    /// Block is only announced on the local network.
+    Local,
+    /// Block is announced.
+    Public,
+}
+
+/// Implementable by ipld storage backends.
+pub trait Store {
+    /// Returns a block from the store. If the block is not in the
+    /// store it fetches it from the network and pins the block. This
+    /// future should be wrapped in a timeout. Dropping the future
+    /// cancels the request.
+    fn get(&self, cid: &Cid) -> StoreResult<Box<[u8]>>;
+
+    /// Inserts and pins a block into the store and announces the block
+    /// if it is visible.
+    fn insert(&self, cid: &Cid, data: Box<[u8]>, visibility: Visibility) -> StoreResult<()>;
+
+    /// Flushes the write buffer.
+    fn flush(&self) -> StoreResult<()>;
+
+    /// Marks a block ready for garbage collection.
+    fn unpin(&self, cid: &Cid) -> StoreResult<()>;
+
+    /// Garbage collects unused blocks in the background.
+    fn gc(&self) -> StoreResult<()>;
```

Comment:
Makes sense but I'm not sure if we want `rust-ipld` to care about pinning or gc? Those two probably deserve a second glance to see if they might actually be better off living in an `ipfs-****` implementation somehow

Reply:
Unpin is essential for applications. I guess gc is less of an issue as ipld-collections or unixfs probably don't care when to gc.

Reply:
I think we can remove `gc` but as I understand you're doing an implementation of this trait currently so I'll wait to see if you end up needing `rust-ipld` to know about it

Reply:
You're right, I think it's better to remove it and have the implementation schedule gc when it thinks it's best